applications/nrf_desktop/src/hw_interface/click.c

```diff
@@ -0,0 +1,234 @@
+	const struct click_detector_config *config;
+	struct k_delayed_work single_click;
+	struct k_delayed_work long_click;
+	bool button_pressed_on_init;
+	u32_t timestamp;
+};
+
+static struct click_detector detectors[ARRAY_SIZE(click_configs)];
+static enum state state;
+
+struct k_delayed_work click_on_init;
+
+static void submit_click_event(struct click_detector *detector,
+			       enum click click)
+{
+	struct click_event *event = new_click_event();
+
+	event->detector_id = detector->config->detector_id;
+	event->click = click;
+
+	EVENT_SUBMIT(event);
+}
+
+static void single_click_handler(struct k_work *work)
+{
+	struct click_detector *detector = CONTAINER_OF(work,
+						       struct click_detector,
+						       single_click);
+	submit_click_event(detector, CLICK_SHORT);
+
+	/* Reset workqueue to make sure cancel will notice work was done. */
+	detector->single_click.work_q = NULL;
+}
+
+static void long_click_handler(struct k_work *work)
+{
+	struct click_detector *detector = CONTAINER_OF(work,
+						       struct click_detector,
+						       long_click);
+	detector->timestamp = 0;
+	submit_click_event(detector, CLICK_LONG);
+
+	/* Reset workqueue to make sure cancel will notice work was done. */
+	detector->single_click.work_q = NULL;
+}
+
+static void click_on_init_handler(struct k_work *work)
+{
+	switch (state) {
+	case STATE_INIT_CLICK_CHECK:
+	{
+		bool click_active = false;
+
+		for (size_t i = 0; i < ARRAY_SIZE(detectors); i++) {
+			if (detectors[i].button_pressed_on_init) {
+				click_active = true;
+				break;
+			}
+		}
+		if (click_active) {
+			k_delayed_work_submit(&click_on_init, HOLD_ON_INIT_MIN);
+			state = STATE_INIT_HOLD_CHECK;
+		} else {
+			state = STATE_ACTIVE;
+		}
+		break;
+	}
+	case STATE_INIT_HOLD_CHECK:
+		for (size_t i = 0; i < ARRAY_SIZE(detectors); i++) {
+			if (detectors[i].button_pressed_on_init) {
+				submit_click_event(&detectors[i],
+						   CLICK_HOLD_ON_INIT);
+			}
+		}
+		state = STATE_ACTIVE;
+		break;
+
+	default:
+		/* Should not happen. */
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+}
+
+static int init(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(detectors); i++) {
+		detectors[i].config = &click_configs[i];
+		k_delayed_work_init(&detectors[i].single_click,
+				    single_click_handler);
+		k_delayed_work_init(&detectors[i].long_click,
+				    long_click_handler);
+	}
+	k_delayed_work_init(&click_on_init, click_on_init_handler);
+	k_delayed_work_submit(&click_on_init, CLICK_ON_INIT_MAX);
+
+	state = STATE_INIT_CLICK_CHECK;
+
+	return 0;
+}
+
+static enum click get_click(u32_t time_diff)
+{
+	if (time_diff < SHORT_CLICK_MAX) {
+		return CLICK_SHORT;
+	}
+	return CLICK_NONE;
+}
+
+static void process_button(struct click_detector *detector, bool pressed)
+{
+	u32_t cur_time = k_uptime_get_32();
+
+	if (pressed) {
+		detector->timestamp = cur_time;
+		k_delayed_work_submit(&detector->long_click, LONG_CLICK_MIN);
+	} else {
+		int err = k_delayed_work_cancel(&detector->single_click);
```

Comment:
according to documentation: The result of calling this on a k_delayed_work item that has not been submitted (i.e. before the return of the k_delayed_work_submit_to_queue() call) is undefined.: https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html?highlight=cancel%20work#_CPPv321k_delayed_work_cancelP14k_delayed_work

Reply:
I checked in code, it would return `-EINVAL`, since` work->work_q` is already initialized to `NULL` on` k_delayed_work_init` (so everything should work fine)

Reply:
The question here is, should we base on internals of implementation - that are volatile, since documentation gives that liberty for zephyr developers? What is that changes in future?

Reply:
@ppryga this doc entry was added after discussion between Piotr and Andy and is there only to close it :) The problem that was discussed was regarding the simultaneous, asynch trigger of cancel and submit. As all our events come from workqueue thread we are not asynch and calling cancel is safe.

Reply:
ok, then part of code is valid.